src/manager/node.c

```diff
@@ -1394,104 +1506,112 @@ static void node_send_agent_subscribe_all(Node *node) {
                         hirte_log_error("Failed to subscribe to unit, OOM");
                         return;
                 }
-                usubs = hashmap_set(node->unit_subscriptions, &v);
-                if (usubs == NULL && hashmap_oom(node->unit_subscriptions)) {
-                        free(v.unit);
-                        hirte_log_error("Failed to subscribe to unit, OOM");
-                        return;
-                }
-
-                /* First sub to this unit, pass to agent */
-                node_send_agent_subscribe(node, sub->unit);
+                usub->sub = sub;
 
                 usubs = hashmap_get(node->unit_subscriptions, &key);
-        }
+                if (usubs == NULL) {
+                        UnitSubscriptions v = { NULL, NULL, false, _UNIT_ACTIVE_STATE_INVALID, NULL };
+                        v.unit = strdup(key.unit);
+                        if (v.unit == NULL) {
+                                hirte_log_error("Failed to subscribe to unit, OOM");
+                                return;
+                        }
 
-        LIST_APPEND(subs, usubs->subs, steal_pointer(&usub));
+                        usubs = hashmap_set(node->unit_subscriptions, &v);
+                        if (usubs == NULL && hashmap_oom(node->unit_subscriptions)) {
+                                free(v.unit);
+                                hirte_log_error("Failed to subscribe to unit, OOM");
+                                return;
+                        }
 
-        /* We know this is loaded, so we won't get notified from
-           the agent, instead send a virtual event here. */
-        if (usubs->loaded) {
-                int r = sub->handle_unit_new(sub->monitor, node->name, sub->unit, "virtual");
-                if (r < 0) {
-                        hirte_log_error("Failed to emit UnitNew signal");
+                        /* First sub to this unit, pass to agent */
+                        node_send_agent_subscribe(node, sub_unit->name);
+
+                        usubs = hashmap_get(node->unit_subscriptions, &key);
                 }
 
-                if (usubs->active_state >= 0) {
-                        r = sub->handle_unit_state_changed(
-                                        sub->monitor,
-                                        node->name,
-                                        sub->unit,
-                                        active_state_to_string(usubs->active_state),
-                                        usubs->substate ? usubs->substate : "invalid",
-                                        "virtual");
+                LIST_APPEND(subs, usubs->subs, steal_pointer(&usub));
+
+                /* We know this is loaded, so we won't get notified from
+                   the agent, instead send a virtual event here. */
+                if (usubs->loaded) {
+                        int r = sub->handle_unit_new(sub->monitor, node->name, sub_unit->name, "virtual");
```

Comment:
This is also potentially sending the same "new" unit twice if you subscribe to the same unit twice. I wonder if we should track that we sent it? This is kinda fringe semantics though.